src/EditorFeatures/Core/Wrapping/SeparatedSyntaxList/SeparatedSyntaxListCodeActionComputer.cs

```diff
@@ -0,0 +1,376 @@
+                result.Add(await GetWrapLongGroupAsync().ConfigureAwait(false));
+            }
+
+            #region unwrap all
+
+            private async Task<WrappingGroup> GetUnwrapGroupAsync()
+            {
+                var unwrapActions = ArrayBuilder<WrapItemsAction>.GetInstance();
+
+                var parentTitle = string.Format(FeaturesResources.Unwrap_0, Wrapper.ListName);
+
+                // MethodName(int a, int b, int c, int d, int e, int f, int g, int h, int i, int j)
+                unwrapActions.Add(await GetUnwrapAllCodeActionAsync(parentTitle, WrappingStyle.UnwrapFirst_IndentRest).ConfigureAwait(false));
+                
+                // MethodName(
+                //      int a, int b, int c, int d, int e, int f, int g, int h, int i, int j)
+                unwrapActions.Add(await GetUnwrapAllCodeActionAsync(parentTitle, WrappingStyle.WrapFirst_IndentRest).ConfigureAwait(false));
+
+                // The 'unwrap' title strings are unique and do not collide with any other code
+                // actions we're computing.  So they can be inlined if possible.
+                return new WrappingGroup(isInlinable: true, unwrapActions.ToImmutableAndFree());
+            }
+
+            private async Task<WrapItemsAction> GetUnwrapAllCodeActionAsync(string parentTitle, WrappingStyle wrappingStyle)
+            {
+                var edits = GetUnwrapAllEdits(wrappingStyle);
+                var title = wrappingStyle == WrappingStyle.WrapFirst_IndentRest
+                    ? string.Format(FeaturesResources.Unwrap_and_indent_all_0, Wrapper.ItemNamePlural)
+                    : string.Format(FeaturesResources.Unwrap_all_0, Wrapper.ItemNamePlural);
+                
+                return await TryCreateCodeActionAsync(edits, parentTitle, title).ConfigureAwait(false);
+            }
+
+            private ImmutableArray<Edit> GetUnwrapAllEdits(WrappingStyle wrappingStyle)
+            {
+                var result = ArrayBuilder<Edit>.GetInstance();
+
+                AddTextChangeBetweenOpenAndFirstItem(wrappingStyle, result);
+
+                foreach (var comma in _listItems.GetSeparators())
+                {
+                    result.Add(Edit.DeleteBetween(comma.GetPreviousToken(), comma));
+                    result.Add(Edit.DeleteBetween(comma, comma.GetNextToken()));
+                }
+
+                result.Add(Edit.DeleteBetween(_listItems.Last(), _listSyntax.GetLastToken()));
+                return result.ToImmutableAndFree();
+            }
+
+            #endregion
+
+            #region wrap long line
+
+            private async Task<WrappingGroup> GetWrapLongGroupAsync()
+            {
+                var parentTitle = string.Format(FeaturesResources.Wrap_long_0, Wrapper.ListName);
+                var codeActions = ArrayBuilder<WrapItemsAction>.GetInstance();
+
+                // MethodName(int a, int b, int c,
```

Comment:
every wrapping mentioned here looks like can be done by formatting engine with custom rules, even for long line wrapping.

all one needs to do is creating custom rule with forcing specific tokens to new line, and adding indentation span for those tokens to go and alignment tokens to align those.

Reply:
I tried, but could not make it work.  I tried debugging through, but the formatting engine was too compelx, with too many non-local decisions being made.  i gave up as it was just much simpler to hand-write this since there was no other logic operations coming in and interfering.  

Now, if we want to support things like wrapping multi-line constructs, or dealing with leading/trailing trivia, i'd be happy if the formatting engine could handle that.  I'd be happy to accept help from you if you want to refactor this to use the formatting engine.  The tests will be good to ensure that nothing breaks.   Since you know the formatting engine the best, and you understand all the quirks it has, you would likely be the best person to take on that work if you're interested.  Thanks!